라이브러리 불러오기

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime

파일 불러오기

In [32]:
train_df = pd.read_csv("smart_farm_train.csv")

In [33]:
train_df.isnull().sum()

시간                 0
년월일                0
년도                 0
월                  0
일                  0
시                  0
분                  0
일사량(W/㎡)           0
외기온(℃)             0
외기습도(%)            0
풍속(m/s)            0
풍향(º)              0
내부 온도            497
내부 습도            497
차광스크린(수평)        497
차열스크린(수평)        497
에너지스크린(수평)       497
에너지스크린(수직)       497
배기팬              497
천창               497
유동팬              497
팬코일 B동           497
환기온도(천창 제어온도)    485
난방온도(설정온도)       484
열공급량(kWh)        915
dtype: int64

In [34]:
train_df.head()

시간                  년월일    년도   월   일  시  분  일사량(W/㎡)  \
0  2021-11-26 00:00:00  2021-11-26 00:00:00  2021  11  26  0  0      -2.7   
1  2021-11-26 00:01:00  2021-11-26 00:00:00  2021  11  26  0  1      -2.9   
2  2021-11-26 00:02:00  2021-11-26 00:00:00  2021  11  26  0  2      -3.0   
3  2021-11-26 00:03:00  2021-11-26 00:00:00  2021  11  26  0  3      -3.0   
4  2021-11-26 00:04:00  2021-11-26 00:00:00  2021  11  26  0  4      -2.8   

   외기온(℃)  외기습도(%)  ...  차열스크린(수평)  에너지스크린(수평)  에너지스크린(수직)  배기팬   천창  유동팬  \
0     5.8       67  ...        0.0         0.0         0.0  0.0  0.0  1.0   
1     5.8       67  ...        0.0         0.0         0.0  0.0  0.0  1.0   
2     5.7       67  ...        0.0         0.0         0.0  0.0  0.0  1.0   
3     5.7       67  ...        0.0         0.0         0.0  0.0  0.0  1.0   
4     5.7       67  ...        0.0         0.0         0.0  0.0  0.0  1.0   

   팬코일 B동  환기온도(천창 제어온도)  난방온도(설정온도)  열공급량(kWh)  
0     0.0           28.0        18.0        0.9  
1     0.0           28.0        18.0        0.9  
2     1.0           28.0        18.0        0.9  
3     1.0           28.0        18.0        0.8  
4     1.0           28.0        18.0        1.0  

[5 rows x 25 columns]

### 내부온도 결측에 대한 확인
- 내부온도 497개 정도의 데이터가 결측이 매우 많은 것으로 의심됨
- 18개의 데이터 컬럼 중에서 5개를 제외하고 모두 결측 -> 충분한 데이터가 존재하지 않음
- 데이터를 제거해야 한다

In [38]:
# 내부온도, 내부습도가 null인 데이터 확인하기
train_df[(train_df["내부 온도"].isnull()) & (train_df["내부 습도"].isnull())]

시간                  년월일    년도   월   일   시   분  \
290     2021-11-26 04:50:00  2021-11-26 00:00:00  2021  11  26   4  50   
291     2021-11-26 04:51:00  2021-11-26 00:00:00  2021  11  26   4  51   
292     2021-11-26 04:52:00  2021-11-26 00:00:00  2021  11  26   4  52   
883     2021-11-26 14:43:00  2021-11-26 00:00:00  2021  11  26  14  43   
884     2021-11-26 14:44:00  2021-11-26 00:00:00  2021  11  26  14  44   
...                     ...                  ...   ...  ..  ..  ..  ..   
106851  2022-02-25 04:51:00  2022-02-25 00:00:00  2022   2  25   4  51   
109730  2022-02-27 04:50:00  2022-02-27 00:00:00  2022   2  27   4  50   
109731  2022-02-27 04:51:00  2022-02-27 00:00:00  2022   2  27   4  51   
111170  2022-02-28 04:50:00  2022-02-28 00:00:00  2022   2  28   4  50   
111171  2022-02-28 04:51:00  2022-02-28 00:00:00  2022   2  28   4  51   

        일사량(W/㎡)  외기온(℃)  외기습도(%)  ...  차열스크린(수평)  에너지스크린(수평)  에너지스크린(수직)  \
290         -3.9     2.1       79  ...        NaN         NaN         NaN   
291         -3.7     2.1       79  ...        NaN         NaN         NaN   
292         -3.7     2.1       78  ...        NaN         NaN         NaN   
883        414.0    13.7       20  ...        NaN         NaN         NaN   
884        410.0    13.7       20  ...        NaN         NaN         NaN   
...          ...     ...      ...  ...        ...         ...         ...   
106851      -4.0    -4.3       61  ...        NaN         NaN         NaN   
109730      -3.5     2.2       51  ...        NaN         NaN         NaN   
109731      -3.5     2.2       51  ...        NaN         NaN         NaN   
111170      -3.6    -2.9       64  ...        NaN         NaN         NaN   
111171      -3.8    -3.0       64  ...        NaN         NaN         NaN   

        배기팬  천창  유동팬  팬코일 B동  환기온도(천창 제어온도)  난방온도(설정온도)  열공급량(kWh)  
290     NaN NaN  NaN     NaN            NaN         NaN        1.9  
291     NaN NaN  NaN     NaN            NaN         NaN        1.3  
292     NaN NaN  NaN     NaN            NaN         NaN        0.8  
883     NaN NaN  NaN     NaN            NaN         NaN        0.1  
884     NaN NaN  NaN     NaN            NaN         NaN        0.1  
...     ...  ..  ...     ...            ...         ...        ...  
106851  NaN NaN  NaN     NaN            NaN         NaN        0.4  
109730  NaN NaN  NaN     NaN            NaN         NaN        0.8  
109731  NaN NaN  NaN     NaN            NaN         NaN        0.8  
111170  NaN NaN  NaN     NaN            NaN         NaN        0.4  
111171  NaN NaN  NaN     NaN            NaN         NaN        0.4  

[497 rows x 25 columns]

In [40]:
# Y값 결측치는 drop해주기
train_df =  train_df[train_df["열공급량(kWh)"].notnull()].reset_index(drop=True)

해당 데이터 제거하기
- 환기온도, 난방온도 아직 결측치 존재

In [41]:
# 내부온도가 null인 512개 데이터 제거하고 다시한번 null 확인
train_df = train_df[train_df["내부 온도"].notnull()].reset_index(drop=True)

train_df.isnull().sum()

시간                 0
년월일                0
년도                 0
월                  0
일                  0
시                  0
분                  0
일사량(W/㎡)           0
외기온(℃)             0
외기습도(%)            0
풍속(m/s)            0
풍향(º)              0
내부 온도              0
내부 습도              0
차광스크린(수평)          0
차열스크린(수평)          0
에너지스크린(수평)         0
에너지스크린(수직)         0
배기팬                0
천창                 0
유동팬                0
팬코일 B동             0
환기온도(천창 제어온도)    133
난방온도(설정온도)       132
열공급량(kWh)          0
dtype: int64

### 결측치 유형
- 환기온도만 결측 존재 : 앞뒤 평균으로 결측치 보충
- 난방온도만 결측 존재 : 앞뒤 평균으로 결측치 보충
- 환기온도, 난방온도 모두 결측 존재 : 회귀기반 예측

In [46]:
train_df[(train_df["환기온도(천창 제어온도)"].isnull()) & (train_df["난방온도(설정온도)"].notnull())].head() # 환기온도만

시간                  년월일    년도   월  일   시   분  일사량(W/㎡)  \
8276  2021-12-01 20:24:00  2021-12-01 00:00:00  2021  12  1  20  24      -3.2   

      외기온(℃)  외기습도(%)  ...  차열스크린(수평)  에너지스크린(수평)  에너지스크린(수직)  배기팬   천창  유동팬  \
8276     1.8       46  ...        0.0         0.0         0.0  0.0  0.0  1.0   

      팬코일 B동  환기온도(천창 제어온도)  난방온도(설정온도)  열공급량(kWh)  
8276     1.0            NaN        15.4        0.9  

[1 rows x 25 columns]

In [50]:
train_df.iloc[8275:8278,:]

시간                  년월일    년도   월  일   시   분  일사량(W/㎡)  \
8275  2021-12-01 20:23:00  2021-12-01 00:00:00  2021  12  1  20  23      -3.2   
8276  2021-12-01 20:24:00  2021-12-01 00:00:00  2021  12  1  20  24      -3.2   
8277  2021-12-01 20:25:00  2021-12-01 00:00:00  2021  12  1  20  25      -3.1   

      외기온(℃)  외기습도(%)  ...  차열스크린(수평)  에너지스크린(수평)  에너지스크린(수직)  배기팬   천창  유동팬  \
8275     1.8       46  ...        0.0         0.0         0.0  0.0  0.0  1.0   
8276     1.8       46  ...        0.0         0.0         0.0  0.0  0.0  1.0   
8277     1.9       46  ...        0.0         0.0         0.0  0.0  0.0  1.0   

      팬코일 B동  환기온도(천창 제어온도)  난방온도(설정온도)  열공급량(kWh)  
8275     1.0            NaN         NaN        1.2  
8276     1.0            NaN        15.4        0.9  
8277     1.0           28.0        15.4        0.1  

[3 rows x 25 columns]

In [48]:
train_df[(train_df["환기온도(천창 제어온도)"].isnull()) & (train_df["난방온도(설정온도)"].isnull())].head() # 환기온도, 난방온도 결측치

시간                  년월일    년도   월   일   시   분  \
5283  2021-11-29 17:34:00  2021-11-29 00:00:00  2021  11  29  17  34   
5493  2021-11-29 21:05:00  2021-11-29 00:00:00  2021  11  29  21   5   
5513  2021-11-29 21:27:00  2021-11-29 00:00:00  2021  11  29  21  27   
5514  2021-11-29 21:30:00  2021-11-29 00:00:00  2021  11  29  21  30   
5515  2021-11-29 21:33:00  2021-11-29 00:00:00  2021  11  29  21  33   

      일사량(W/㎡)  외기온(℃)  외기습도(%)  ...  차열스크린(수평)  에너지스크린(수평)  에너지스크린(수직)  배기팬  \
5283      -2.7    13.4       44  ...      100.0         0.0         0.0  0.0   
5493      -2.8     6.3       72  ...        0.0         0.0         0.0  0.0   
5513      -3.5     6.1       73  ...        0.0         0.0         0.0  0.0   
5514      -3.6     6.0       73  ...        0.0         0.0         0.0  0.0   
5515      -3.5     5.8       73  ...        0.0         0.0         0.0  0.0   

       천창  유동팬  팬코일 B동  환기온도(천창 제어온도)  난방온도(설정온도)  열공급량(kWh)  
5283  0.0  1.0     0.0            NaN         NaN        0.0  
5493  0.0  1.0     1.0            NaN         NaN        1.5  
5513  0.0  1.0     1.0            NaN         NaN        0.0  
5514  0.0  1.0     0.0            NaN         NaN        0.0  
5515  0.0  1.0     0.0            NaN         NaN        0.0  

[5 rows x 25 columns]

#### 앞뒤 데이터 평균으로 결측치 보간하기

In [52]:
def get_온도_결측치(data):
    
    측정목록 = ["환기온도(천창 제어온도)","난방온도(설정온도)"]
    결과리스트 = []

    # 데이터프레임 copy하기
    df = data.copy()
    
    for 항목 in 측정목록:
        for idx,row in df[df[항목].isnull()].iterrows():
            
            if df.loc[idx+1,항목] != np.nan:
                
                if df.loc[idx-1,항목] != np.nan:
                    
                    df.loc[idx,항목] = ( df.loc[idx+1,항목]+df.loc[idx-1,항목] ) / 2
                    
    결과리스트.append(df)
    
    return pd.concat(결과리스트,axis=0,ignore_index = True)

In [53]:
train_no_df = get_온도_결측치(train_df)

In [54]:
print("결측치 처리 이전",train_df.isnull().sum().sum())
print("결측치 처리 이후",train_no_df.isnull().sum().sum())

결측치 처리 이전 265
결측치 처리 이후 227


In [ ]:
smart_farm_null
smart_farm_nonull

#### 회귀대체 방법
- 2개 열 모두 결측인 케이스 : 113개
- 1개 열만 결측인 케이스 : 1개

In [66]:
train_no_df[(train_no_df["환기온도(천창 제어온도)"].isnull()) & (train_no_df["난방온도(설정온도)"].isnull())].shape

(113, 25)

In [69]:
train_no_df[(train_no_df["환기온도(천창 제어온도)"].isnull()) & (train_no_df["난방온도(설정온도)"].notnull())]

시간                  년월일    년도   월  일   시   분  일사량(W/㎡)  \
8276  2021-12-01 20:24:00  2021-12-01 00:00:00  2021  12  1  20  24      -3.2   

      외기온(℃)  외기습도(%)  ...  차열스크린(수평)  에너지스크린(수평)  에너지스크린(수직)  배기팬   천창  유동팬  \
8276     1.8       46  ...        0.0         0.0         0.0  0.0  0.0  1.0   

      팬코일 B동  환기온도(천창 제어온도)  난방온도(설정온도)  열공급량(kWh)  
8276     1.0            NaN        15.4        0.9  

[1 rows x 25 columns]

#### 1개 열만 결측인 경우
- train_all_no_df : 평균 보간 이후 결측치가 아예 없는 것

In [106]:
from sklearn import linear_model

In [107]:
# 결측치 전부 제거
train_all_no_df = train_no_df.dropna(axis=0,how="any")

vent_temp = train_no_df[train_no_df["난방온도(설정온도)"].notnull()]
# vent_temp(환기 온도) : 난방온도(천창 제어온도)가 null이 아닌 데이터

In [108]:
vent_temp[vent_temp["환기온도(천창 제어온도)"].isnull()]

Empty DataFrame
Columns: [시간, 년월일, 년도, 월, 일, 시, 분, 일사량(W/㎡), 외기온(℃), 외기습도(%), 풍속(m/s), 풍향(º), 내부 온도, 내부 습도, 차광스크린(수평), 차열스크린(수평), 에너지스크린(수평), 에너지스크린(수직), 배기팬, 천창, 유동팬, 팬코일 B동, 환기온도(천창 제어온도), 난방온도(설정온도), 열공급량(kWh)]
Index: []

[0 rows x 25 columns]

vent_temp : 환기온도 채우기

In [109]:
# 환기온도를 제외한 나머지 컬럼으로 예측하기

col_lst = ['일사량(W/㎡)', '외기온(℃)', '외기습도(%)',
       '풍속(m/s)', '풍향(º)', '내부 온도', '내부 습도', '차광스크린(수평)', '차열스크린(수평)',
       '에너지스크린(수평)', '에너지스크린(수직)', '배기팬', '천창', '유동팬', '팬코일 B동',
                   '난방온도(설정온도)', '열공급량(kWh)']

X = train_all_no_df[col_lst]

y = train_all_no_df["환기온도(천창 제어온도)"]

In [103]:
reg = linear_model.LinearRegression()
reg.fit(X,y)

y_pred = reg.predict(vent_temp[col_lst])

# 모든 행의 환기온도를 예측한 데이터
vent_temp_pred = pd.DataFrame(vent_temp[['시간','환기온도(천창 제어온도)']].fillna(pd.Series(y_pred.flatten())))

# 기존의 환기온도 데이터에서 결측 부분에서 예측한 데이터를 채워줌
vent_temp['환기온도(천창 제어온도)'].fillna(pd.Series(y_pred.flatten()), inplace=True)

In [110]:
# 예측한 값들 채우기
train_no_df.loc[vent_temp.index,"환기온도(천창 제어온도)"] = train_no_df.loc[vent_temp.index,
                                                   "환기온도(천창 제어온도)"].fillna(vent_temp["환기온도(천창 제어온도)"])

In [117]:
vent_temp.loc[8274:8278,:]

시간                  년월일    년도   월  일   시   분  일사량(W/㎡)  \
8276  2021-12-01 20:24:00  2021-12-01 00:00:00  2021  12  1  20  24      -3.2   
8277  2021-12-01 20:25:00  2021-12-01 00:00:00  2021  12  1  20  25      -3.1   
8278  2021-12-01 20:26:00  2021-12-01 00:00:00  2021  12  1  20  26      -3.3   

      외기온(℃)  외기습도(%)  ...  차열스크린(수평)  에너지스크린(수평)  에너지스크린(수직)  배기팬   천창  유동팬  \
8276     1.8       46  ...        0.0         0.0         0.0  0.0  0.0  1.0   
8277     1.9       46  ...        0.0         0.0         0.0  0.0  0.0  1.0   
8278     1.9       46  ...        0.0         0.0         0.0  0.0  0.0  1.0   

      팬코일 B동  환기온도(천창 제어온도)  난방온도(설정온도)  열공급량(kWh)  
8276     1.0      28.133389        15.4        0.9  
8277     1.0      28.000000        15.4        0.1  
8278     1.0      28.000000        15.5        1.0  

[3 rows x 25 columns]

#### 2개 열만 결측인 경우
- smart_farm_all_nonull : 평균 보간 이후 결측치가 아예 없는 것
- smart_farm_nonull : 결측치가 1개열만 있는 것은 채워지고 2개 열에 대한 결측만 존재하는 상태

index 이상없게 reset 한번 해주기

In [119]:
train_no_df.reset_index(drop=True,inplace=True)

In [120]:
# 환기온도, 난방온도 아예 제외
col_lst = ['일사량(W/㎡)', '외기온(℃)', '외기습도(%)',
       '풍속(m/s)', '풍향(º)', '내부 온도', '내부 습도', '차광스크린(수평)', '차열스크린(수평)',
       '에너지스크린(수평)', '에너지스크린(수직)', '배기팬', '천창', '유동팬', '팬코일 B동', '열공급량(kWh)']

heat_vent_temp = train_no_df[(train_no_df["난방온도(설정온도)"].isnull()) &
                                   (train_no_df["환기온도(천창 제어온도)"].isnull())]

In [126]:
heat_vent_temp.head()

시간                  년월일    년도   월   일   시   분  \
5513  2021-11-29 21:27:00  2021-11-29 00:00:00  2021  11  29  21  27   
5514  2021-11-29 21:30:00  2021-11-29 00:00:00  2021  11  29  21  30   
5515  2021-11-29 21:33:00  2021-11-29 00:00:00  2021  11  29  21  33   
5516  2021-11-29 21:36:00  2021-11-29 00:00:00  2021  11  29  21  36   
5517  2021-11-29 21:39:00  2021-11-29 00:00:00  2021  11  29  21  39   

      일사량(W/㎡)  외기온(℃)  외기습도(%)  ...  차열스크린(수평)  에너지스크린(수평)  에너지스크린(수직)  배기팬  \
5513      -3.5     6.1       73  ...        0.0         0.0         0.0  0.0   
5514      -3.6     6.0       73  ...        0.0         0.0         0.0  0.0   
5515      -3.5     5.8       73  ...        0.0         0.0         0.0  0.0   
5516      -3.7     5.7       73  ...        0.0         0.0         0.0  0.0   
5517      -3.6     5.5       74  ...        0.0         0.0         0.0  0.0   

       천창  유동팬  팬코일 B동  환기온도(천창 제어온도)  난방온도(설정온도)  열공급량(kWh)  
5513  0.0  1.0     1.0      27.980521   17.757187        0.0  
5514  0.0  1.0     0.0      28.007929   17.240702        0.0  
5515  0.0  1.0     0.0      27.936869   17.109440        0.0  
5516  0.0  1.0     1.0      27.975362   17.653200        0.2  
5517  0.0  1.0     1.0      28.008288   17.818466        0.6  

[5 rows x 25 columns]

In [127]:
X = train_all_no_df[col_lst]

y_1 = train_all_no_df["난방온도(설정온도)"] ; y_2 = train_all_no_df["환기온도(천창 제어온도)"]

# 각각 예측하기
reg_1 = linear_model.LinearRegression() ; reg_1.fit(X,y_1)
reg_2 = linear_model.LinearRegression() ; reg_2.fit(X,y_2)

y_pred_1 = reg_1.predict(heat_vent_temp[col_lst])
y_pred_2 = reg_2.predict(heat_vent_temp[col_lst])

heat_vent_temp['난방온도(설정온도)'] =  y_pred_1.flatten()
heat_vent_temp['환기온도(천창 제어온도)'] = y_pred_2.flatten()

In [128]:
heat_vent_temp.head()

시간                  년월일    년도   월   일   시   분  \
5513  2021-11-29 21:27:00  2021-11-29 00:00:00  2021  11  29  21  27   
5514  2021-11-29 21:30:00  2021-11-29 00:00:00  2021  11  29  21  30   
5515  2021-11-29 21:33:00  2021-11-29 00:00:00  2021  11  29  21  33   
5516  2021-11-29 21:36:00  2021-11-29 00:00:00  2021  11  29  21  36   
5517  2021-11-29 21:39:00  2021-11-29 00:00:00  2021  11  29  21  39   

      일사량(W/㎡)  외기온(℃)  외기습도(%)  ...  차열스크린(수평)  에너지스크린(수평)  에너지스크린(수직)  배기팬  \
5513      -3.5     6.1       73  ...        0.0         0.0         0.0  0.0   
5514      -3.6     6.0       73  ...        0.0         0.0         0.0  0.0   
5515      -3.5     5.8       73  ...        0.0         0.0         0.0  0.0   
5516      -3.7     5.7       73  ...        0.0         0.0         0.0  0.0   
5517      -3.6     5.5       74  ...        0.0         0.0         0.0  0.0   

       천창  유동팬  팬코일 B동  환기온도(천창 제어온도)  난방온도(설정온도)  열공급량(kWh)  
5513  0.0  1.0     1.0      27.980521   17.757187        0.0  
5514  0.0  1.0     0.0      28.007929   17.240702        0.0  
5515  0.0  1.0     0.0      27.936869   17.109440        0.0  
5516  0.0  1.0     1.0      27.975362   17.653200        0.2  
5517  0.0  1.0     1.0      28.008288   17.818466        0.6  

[5 rows x 25 columns]

index 추출 5513,5514,5515,5516...

train_no_df의 5513,5514,5515,5516... 결측치를

heat_vent_temp의 5513,5514,5515,5516... 결측치 해결한 값으로 채우기

In [129]:
train_no_df.loc[heat_vent_temp.index,"난방온도(설정온도)"] = train_no_df.loc[heat_vent_temp.index,
                                                   "난방온도(설정온도)"].fillna(heat_vent_temp["난방온도(설정온도)"])


train_no_df.loc[heat_vent_temp.index,"환기온도(천창 제어온도)"] = train_no_df.loc[heat_vent_temp.index,
                                                   "환기온도(천창 제어온도)"].fillna(heat_vent_temp["환기온도(천창 제어온도)"])

In [130]:
train_no_df.loc[[5513,5514],:]

시간                  년월일    년도   월   일   시   분  \
5513  2021-11-29 21:27:00  2021-11-29 00:00:00  2021  11  29  21  27   
5514  2021-11-29 21:30:00  2021-11-29 00:00:00  2021  11  29  21  30   

      일사량(W/㎡)  외기온(℃)  외기습도(%)  ...  차열스크린(수평)  에너지스크린(수평)  에너지스크린(수직)  배기팬  \
5513      -3.5     6.1       73  ...        0.0         0.0         0.0  0.0   
5514      -3.6     6.0       73  ...        0.0         0.0         0.0  0.0   

       천창  유동팬  팬코일 B동  환기온도(천창 제어온도)  난방온도(설정온도)  열공급량(kWh)  
5513  0.0  1.0     1.0      27.980521   17.757187        0.0  
5514  0.0  1.0     0.0      28.007929   17.240702        0.0  

[2 rows x 25 columns]